bug: MongoDbVectorIndex - .dynamic_context(64, index) compile error - expected `Value`, found `Bson`

- [X] I have looked for existing issues (including closed) about this

## Bug Report

I am trying to use a MongoDbVectorIndex as a dynamic context index.

It does not compile and fails with

```scala
error[E0271]: type mismatch resolving `<MongoDbSearchFilter as SearchFilter>::Value == Value`
   --> src/main.rs:217:30
    |
217 |         .dynamic_context(64, index)
    |          ---------------     ^^^^^ expected `Value`, found `Bson`
    |          |
    |          required by a bound introduced by this call
    |
    = note: required for `MongoDbVectorIndex<mongodb::bson::Document, rig::providers::ollama::EmbeddingModel<ReqwestClient>>` to implement `VectorStoreIndexDyn`
```

## Reproduction

inspired from [rig-integrations/rig-mongodb/examples/vector_search_mongodb.rs](https://github.com/0xPlaygrounds/rig/blob/main/rig-integrations/rig-mongodb/examples/vector_search_mongodb.rs)

```scala
let index: MongoDbVectorIndex<Document, EmbeddingModel<Client>> =
        MongoDbVectorIndex::new(
            collection,
            embeddings_model,
            "vector_index",
            SearchParams::new(),
        )
        .unwrap();

let agent = AgentBuilder::new(model)
        .preamble("You are a not compiling unhelpful assistant.")
        .dynamic_context(64, index)
        .build();
```

Very possible I am doing it incorrectly.

Dependencies:
```toml
rig-core = { version = "0.29.0", features = ["pdf", "rmcp", "derive"] }
rig-mongodb = "0.3.5"
serde = { version = "1.0.228", features = ["derive"] }
mongodb = "3.5.0"
serde_json = "1.0.149"
reqwest = "0.13.1"
```

--- a/rig-integrations/rig-milvus/src/filter.rs
+++ b/rig-integrations/rig-milvus/src/filter.rs
@@ -76,11 +76,9 @@ impl TryFrom<serde_json::Value> for MilvusValue {
                     .map(MilvusValue::try_from)
                     .collect::<Result<_, _>>()?,
             )),
-            serde_json::Value::Null | serde_json::Value::Object(_) => {
-                Err(FilterError::TypeError(
-                    "Milvus filter does not currently support null values or obejcts".into(),
-                ));
-            }
+            serde_json::Value::Null | serde_json::Value::Object(_) => Err(FilterError::TypeError(
+                "Milvus filter does not currently support null values or objects".into(),
+            )),
         }
     }
 }

--- a/rig-integrations/rig-milvus/src/filter.rs
+++ b/rig-integrations/rig-milvus/src/filter.rs
@@ -1,4 +1,4 @@
-use rig::vector_store::request::SearchFilter;
+use rig::vector_store::request::{Filter as CoreFilter, FilterError, SearchFilter};
 use serde::{Deserialize, Serialize};
 
 pub enum MilvusValue {
@@ -54,6 +54,34 @@ impl From<bool> for MilvusValue {
 impl From<String> for MilvusValue {
     fn from(value: String) -> Self {
         Self::String(value)
+    }
+}
+
+impl TryFrom<serde_json::Value> for MilvusValue {
+    type Error = FilterError;
+    fn try_from(value: serde_json::Value) -> Result<Self, Self::Error> {
+        match value {
+            serde_json::Value::Bool(b) => Ok(MilvusValue::Bool(b)),
+            serde_json::Value::Number(n) => {
+                Ok(MilvusValue::Number(n.as_f64().ok_or_else(|| {
+                    FilterError::Expected {
+                        expected: "Valid 64-bit float".into(),
+                        got: "Invalid 64-bit float".into(),
+                    }
+                })?))
+            }
+            serde_json::Value::String(s) => Ok(MilvusValue::String(s)),
+            serde_json::Value::Array(arr) => Ok(MilvusValue::Array(
+                arr.into_iter()
+                    .map(MilvusValue::try_from)
+                    .collect::<Result<_, _>>()?,
+            )),
+            serde_json::Value::Null | serde_json::Value::Object(_) => {
+                Err(FilterError::TypeError(
+                    "Milvus filter does not currently support null values or obejcts".into(),
+                ));
+            }
+        }
     }
 }
 
@@ -183,5 +211,20 @@ impl Filter {
 
     pub fn into_inner(self) -> String {
         self.0
+    }
+}
+
+impl TryFrom<CoreFilter<serde_json::Value>> for Filter {
+    type Error = FilterError;
+    fn try_from(value: CoreFilter<serde_json::Value>) -> Result<Self, Self::Error> {
+        let value = match value {
+            CoreFilter::Eq(k, val) => Filter::eq(k, val.try_into()?),
+            CoreFilter::Gt(k, val) => Filter::gt(k, val.try_into()?),
+            CoreFilter::Lt(k, val) => Filter::lt(k, val.try_into()?),
+            CoreFilter::And(l, r) => Self::try_from(*l)?.and(Self::try_from(*r)?),
+            CoreFilter::Or(l, r) => Self::try_from(*l)?.or(Self::try_from(*r)?),
+        };
+
+        Ok(value)
     }
 }

--- a/rig-integrations/rig-milvus/src/lib.rs
+++ b/rig-integrations/rig-milvus/src/lib.rs
@@ -5,9 +5,10 @@ use rig::{
     Embed, OneOrMany,
     embeddings::{Embedding, EmbeddingModel},
     vector_store::{
-        InsertDocuments, VectorStoreError, VectorStoreIndex,
-        request::{SearchFilter, VectorSearchRequest},
+        InsertDocuments, TopNResults, VectorStoreError, VectorStoreIndex, VectorStoreIndexDyn,
+        request::{Filter as CoreFilter, SearchFilter, VectorSearchRequest},
     },
+    wasm_compat::WasmBoxedFuture,
 };
 use serde::{Deserialize, Serialize};
 
@@ -307,5 +308,35 @@ where
             .collect();
 
         Ok(res)
+    }
+}
+
+impl<M> VectorStoreIndexDyn for MilvusVectorStore<M>
+where
+    M: EmbeddingModel + Sync + Send,
+{
+    fn top_n<'a>(
+        &'a self,
+        req: VectorSearchRequest<CoreFilter<serde_json::Value>>,
+    ) -> WasmBoxedFuture<'a, TopNResults> {
+        Box::pin(async move {
+            let req = req.try_map_filter(Filter::try_from)?;
+            let results = <Self as VectorStoreIndex>::top_n::<serde_json::Value>(self, req).await?;
+
+            Ok(results)
+        })
+    }
+
+    /// Implement the `top_n_ids` method of the `VectorStoreIndex` trait for `MongoDbVectorIndex`.
+    fn top_n_ids<'a>(
+        &'a self,
+        req: VectorSearchRequest<CoreFilter<serde_json::Value>>,
+    ) -> WasmBoxedFuture<'a, Result<Vec<(f64, String)>, VectorStoreError>> {
+        Box::pin(async move {
+            let req = req.try_map_filter(Filter::try_from)?;
+            let results = <Self as VectorStoreIndex>::top_n_ids(self, req).await?;
+
+            Ok(results)
+        })
     }
 }

--- a/rig-integrations/rig-mongodb/src/lib.rs
+++ b/rig-integrations/rig-mongodb/src/lib.rs
@@ -1,13 +1,14 @@
 use futures::StreamExt;
-use mongodb::bson::{self, Bson, Document, doc};
+use mongodb::bson::{self, Bson, Document, doc, to_bson};
 
 use rig::{
     Embed, OneOrMany,
     embeddings::embedding::{Embedding, EmbeddingModel},
     vector_store::{
-        InsertDocuments, VectorStoreError, VectorStoreIndex,
-        request::{SearchFilter, VectorSearchRequest},
+        InsertDocuments, TopNResults, VectorStoreError, VectorStoreIndex, VectorStoreIndexDyn,
+        request::{Filter, SearchFilter, VectorSearchRequest},
     },
+    wasm_compat::WasmBoxedFuture,
 };
 use serde::{Deserialize, Serialize};
 
@@ -306,13 +307,38 @@ impl MongoDbSearchFilter {
         Self(doc! { key: { "$size": size } })
     }
 
-    // Array ops
+    // Array opsreq.filter.unwrap().
     pub fn all(key: String, values: Vec<Bson>) -> Self {
         Self(doc! { key: { "$all": values } })
     }
 
     pub fn any(key: String, condition: Document) -> Self {
         Self(doc! { key: { "$elemMatch": condition } })
+    }
+}
+
+impl From<Filter<serde_json::Value>> for MongoDbSearchFilter {
+    fn from(value: Filter<serde_json::Value>) -> Self {
+        fn serde_json_value_to_bson(v: &serde_json::Value) -> Bson {
+            to_bson(v).unwrap_or(Bson::Null)
+        }
+
+        match value {
+            Filter::Eq(k, val) => {
+                let bson_val = serde_json_value_to_bson(&val);
+                MongoDbSearchFilter::eq(k, bson_val)
+            }
+            Filter::Gt(k, val) => {
+                let bson_val = serde_json_value_to_bson(&val);
+                MongoDbSearchFilter::gt(k, bson_val)
+            }
+            Filter::Lt(k, val) => {
+                let bson_val = serde_json_value_to_bson(&val);
+                MongoDbSearchFilter::lt(k, bson_val)
+            }
+            Filter::And(l, r) => Self::from(*l).and(Self::from(*r)),
+            Filter::Or(l, r) => Self::from(*l).or(Self::from(*r)),
+        }
     }
 }
 
@@ -411,6 +437,38 @@ where
         );
 
         Ok(results)
+    }
+}
+
+impl<C, M> VectorStoreIndexDyn for MongoDbVectorIndex<C, M>
+where
+    C: Sync + Send,
+    M: EmbeddingModel + Sync + Send,
+{
+    fn top_n<'a>(
+        &'a self,
+        req: VectorSearchRequest<Filter<serde_json::Value>>,
+    ) -> WasmBoxedFuture<'a, TopNResults> {
+        let req = req.map_filter(MongoDbSearchFilter::from);
+
+        Box::pin(async move {
+            let results = <Self as VectorStoreIndex>::top_n::<serde_json::Value>(self, req).await?;
+
+            Ok(results)
+        })
+    }
+
+    /// Implement the `top_n_ids` method of the `VectorStoreIndex` trait for `MongoDbVectorIndex`.
+    fn top_n_ids<'a>(
+        &'a self,
+        req: VectorSearchRequest<Filter<serde_json::Value>>,
+    ) -> WasmBoxedFuture<'a, Result<Vec<(f64, String)>, VectorStoreError>> {
+        let req = req.map_filter(MongoDbSearchFilter::from);
+        Box::pin(async move {
+            let results = <Self as VectorStoreIndex>::top_n_ids(self, req).await?;
+
+            Ok(results)
+        })
     }
 }
 

--- a/rig/rig-core/src/vector_store/request.rs
+++ b/rig/rig-core/src/vector_store/request.rs
@@ -68,6 +68,24 @@ impl<Filter> VectorSearchRequest<Filter> {
             filter: self.filter.map(f),
         }
     }
+
+    /// Transforms the filter type using a provided function which can additionally return a result.
+    ///
+    /// Useful for converting between filter representations where the conversion can potentially fail (eg, unrepresentable or invalid values).
+    pub fn try_map_filter<T, F>(self, f: F) -> Result<VectorSearchRequest<T>, FilterError>
+    where
+        F: Fn(Filter) -> Result<T, FilterError>,
+    {
+        let filter = self.filter.map(f).transpose()?;
+
+        Ok(VectorSearchRequest {
+            query: self.query,
+            samples: self.samples,
+            threshold: self.threshold,
+            additional_params: self.additional_params,
+            filter,
+        })
+    }
 }
 
 /// Errors from constructing or converting filter expressions.

